codec/vtproto/codec.go

```diff
@@ -0,0 +1,47 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package vtproto
+
+import "fmt"
+
+type vtprotoMessage interface {
+	MarshalVT() ([]byte, error)
+	UnmarshalVT([]byte) error
+}
+
+// Codec is a composite of Encoder and Decoder.
+type Codec struct {
+}
+
+// Encode encodes vtprotoMessage type into byte slice
+func (Codec) Encode(in any) ([]byte, error) {
```

Comment:
rather than accept `in any` and then doing a type assertion on top, I think it would be better to the compiler tell us when the passed type isn't right, rather than throw a runtime error.